examples/with-grid/index.js

```diff
@@ -46,6 +49,169 @@ connectButton.onclick = () => {
 
   startSyft(gridServer.value, protocol.value);
 };
+*/
+
+startButton.onclick = () => {
+  setFLUI();
+  startFL(gridServer.value, 'model-id');
+};
+
+const executeFLTrainingJob = async ({
+  data,
+  targets,
+  job,
+  model,
+  clientConfig,
+  callbacks
+}) => {
+  const batchSize = clientConfig.batch_size;
+  const lr = clientConfig.lr;
+  const numBatches = Math.ceil(data.shape[0] / batchSize);
+  const maxEpochs = clientConfig.max_epochs || 1;
+  const maxUpdates = clientConfig.max_updates || maxEpochs * numBatches;
+  // set the lowest cap
+  const numUpdates = Math.min(maxUpdates, maxEpochs * numBatches);
+
+  // Copy original model params.
+  let modelParams = [];
+  for (let param of model.params) {
+    modelParams.push(param.clone());
+  }
+
+  for (let update = 0, batch = 0, epoch = 0; update < numUpdates; update++) {
+    const chunkSize = Math.min(batchSize, data.shape[0] - batch * batchSize);
+    const dataBatch = data.slice(batch * batchSize, chunkSize);
+    const targetBatch = targets.slice(batch * batchSize, chunkSize);
+
+    let [loss, acc, ...newModelParams] = await job.plans[
+      'training_plan'
+    ].execute(
+      job.worker,
+      dataBatch,
+      targetBatch,
+      chunkSize,
+      lr,
+      ...modelParams
+    );
+
+    // Use updated model params in the next cycle.
+    for (let i = 0; i < modelParams.length; i++) {
+      modelParams[i].dispose();
+      modelParams[i] = newModelParams[i];
+    }
+
+    if (typeof callbacks.onBatchEnd === 'function') {
+      callbacks.onBatchEnd({
+        update,
+        batch,
+        epoch,
+        accuracy: (await acc.data())[0],
+        loss: (await loss.data())[0]
+      });
+    }
+
+    batch++;
+    // check if we're out of batches (end of epoch)
+    if (batch === numBatches) {
+      if (typeof callbacks.onEpochEnd === 'function') {
+        callbacks.onEpochEnd({ update, batch, epoch, model });
+      }
+      batch = 0;
+      epoch++;
+    }
+
+    // free GPU memory
+    acc.dispose();
+    loss.dispose();
+    dataBatch.dispose();
+    targetBatch.dispose();
+  }
+
+  // TODO protocol execution
+  // job.protocols['secure_aggregation'].execute();
+
+  // Calc model diffs
+  const modelDiff = [];
+  for (let i = 0; i < modelParams.length; i++) {
+    modelDiff.push(model.params[i].sub(modelParams[i]));
+  }
+
+  // report
+  await job.report(modelDiff);
+
+  if (typeof callbacks.onDone === 'function') {
+    callbacks.onDone();
+  }
+};
+
+const startFL = async (url, modelId) => {
+  const worker = new Syft({ url, verbose: true });
+  const job = await worker.newJob({ modelId });
+  job.start();
+  job.on('ready', async ({ model, clientConfig }) => {
+    // load data
+    console.log('Loading data...');
+    const mnist = new MnistData();
+    await mnist.load();
+    const data = mnist.getTrainData();
+    console.log('Data loaded');
+
+    // train
+    executeFLTrainingJob({
```

Comment:
This works, but I'd prefer to have all the code from this function here, rather than a method call. It's one less thing for people to read and then go "wait, where is that function created?"